File: 0394-decode-string/0394-decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack, num = [], []
        i, n = 0, len(s)
        while i < n:
            exp = []
            if s[i].isdigit():
                if s[i+1].isdigit():
                    digit = s[i]
                    j = i+1
                    while j < n:
                        if s[j].isdigit():
                            digit += s[j]
                            j += 1
                        else:
                            break
                    num.append(int(digit))
                    i = j-1
                else:
                    num.append(int(s[i]))  
            elif s[i] == '[':
                stack.append(s[i])
            elif s[i] == ']':
                while stack and stack[-1]!= "[":
                    exp.append(stack.pop())
                stack.pop()
                exp = "".join(exp[::-1])
                stack.append(exp * num[-1])
                num.pop()
            else:
                stack.append(s[i])
            i +=1
        return "".join(stack)
                     

File: 0394-decode-string/0394-decode-string.java
class Solution {
    public String decodeString(String s) {
        Stack<String> stack = new Stack<>();
        for(int i = 0;i<s.length();i++) {
			Character c = s.charAt(i);
			String current = "";
			if(c == ']') {
				StringBuilder word = new StringBuilder();
				while(!stack.isEmpty() && stack.peek().charAt(0) != '[') 
				{
					String curr = stack.pop();
					if(!curr.equals(current) && Character.isAlphabetic(curr.charAt(0))) 
					{
						word.insert(0,curr);
					}
				}
			
				if(!stack.isEmpty() && stack.peek().charAt(0) == '[') {
					stack.pop();
				}
				
				StringBuilder num = new StringBuilder();
				while(!stack.isEmpty() && Character.isDigit(stack.peek().charAt(0))) 
				{
					num.append(stack.pop());
				}
				num.reverse();
				
				System.out.println(word + " " + num);
				int count = Integer.parseInt(num.toString());
				String x = word.toString();
				for(int j = 0; j < count-1;j++) {
					word.append(x);
				}
				System.out.println("after repeat: "+word);
				stack.add(word.toString());
				System.out.println("Stack: "+stack);
				current = word.toString();
				System.out.println("current: "+current);
				
			}
			else {
				stack.add(c.toString());
			}
		}
		System.out.println("final stack: "+stack);

		StringBuilder result = new StringBuilder();;
		while(!stack.isEmpty()) {
			result.insert(0,stack.pop());
		}
		return result.toString();
    }
}